### DISCUSSION TITLE
Make QueryClientConfig generic

### DISCUSSION THREAD

**Discussion by SevenOutman** [TanStack/query#3528] [Ideas] [answered]
Current declaration for `QueryClientConfig` makes it impossible to specify an `TError` type for `DefaultOptions`.
https://github.com/tannerlinsley/react-query/blob/0c667273125e5307ff148c0ce621b5d24950517f/src/core/types.ts#L679-L688

This makes things hard when your `TError` is certain, e.g. `AxiosError`.

I'd suggest make `QueryClientConfig` generic that accepts a `TError` type argument.

```typescript
export interface QueryClientConfig<TError = unknown> { 
   queryCache?: QueryCache 
   mutationCache?: MutationCache 
   defaultOptions?: DefaultOptions<TError>
 } 
  
 export interface DefaultOptions<TError = unknown> { 
   queries?: QueryObserverOptions<unknown, TError> 
   mutations?: MutationObserverOptions<unknown, TError, unknown, unknown> 
 } 
```

**Comment by TkDodo** [comment, marked as answer]
> This makes things hard when your TError is certain, e.g. AxiosError.

is it though? I don't think you can be certain of that. For example, if you have a runtime error in `select`, your query will also go into `Error` state with that error. A simple `instanceOf Error` check, or to use a runtime type guard from axios (`axios.isAxiosError(error)`) is one line of code.

https://tkdodo.eu/blog/react-query-and-type-script#what-about-error

**Reply by SevenOutman** [reply]
You're right. I didn't realize that `select` error also triggers `onError`. But yes I still missed the point that `queryFn` may throw any runtime error as well. Thanks!

**Reply by SevenOutman** [reply]
Hi @TkDodo, I just couldn't think up why `DefaultOptions` has this `TError` arg then?

**Reply by TkDodo** [reply]
yeah I also don't know because we never use that generic. Maybe we should just remove it?
